Sources/NIO/BaseSocketChannel.swift

```diff
@@ -14,6 +14,163 @@
+    internal mutating func activate(promise: EventLoopPromise<Void>?) -> (() -> Void) {
+        return self.moveState(event: .activate, promise: promise)
+    }
+
+    // MARK: private API
+    @inline(__always) // we need to return a closure here and to not suffer from a potential allocation for that this must be inlined
+    private mutating func moveState(event: Event, promise: EventLoopPromise<Void>?) -> (() -> Void) {
+        assert(self.eventLoop.inEventLoop)
+
+        switch (self.currentState, event) {
+        // origin: .fresh
+        case (.fresh, .register):
+            return self.doStateTransfer(newState: .registered, promise: promise) { pipeline in
+                pipeline.fireChannelRegistered0()
+            }
+
+        case (.fresh, .close):
+            return self.doStateTransfer(newState: .closed, promise: promise) { (_: ChannelPipeline) in }
+
+        // origin: .registered
+        case (.registered, .activate):
+            return self.doStateTransfer(newState: .activated, promise: promise) { pipeline in
+                pipeline.fireChannelActive0()
+            }
+
+        case (.registered, .close):
+            return self.doStateTransfer(newState: .closed, promise: promise) { pipeline in
+                pipeline.fireChannelUnregistered0()
+            }
+
+        // origin: .activated
+        case (.activated, .close):
+            return self.doStateTransfer(newState: .closed, promise: promise) { pipeline in
+                pipeline.fireChannelInactive0()
+                pipeline.fireChannelUnregistered0()
+            }
+
+        // bad transitions
+        case (.fresh, .activate),      // should go through .registered first
+             (.registered, .register), // already registered
+             (.activated, .activate),  // already activated
+             (.activated, .register),  // already registered (and activated)
+             (.closed, _):             // already closed
+            self.badTransition(event: event)
+        }
+    }
+
+    private func badTransition(event: Event) -> Never {
+        fatalError("illegal transition: state=\(self.currentState), event=\(event)")
+    }
+
+    @inline(__always) // we need to return a closure here and to not suffer from a potential allocation for that this must be inlined
+    private mutating func doStateTransfer(newState: State, promise: EventLoopPromise<Void>?, _ callouts: @escaping (ChannelPipeline) -> Void) -> (() -> Void) {
+        self.currentState = newState
+
+        let pipeline = self.channelPipeline
+        return {
+            promise?.succeed(result: ())
+            callouts(pipeline)
+        }
+    }
+
+    // MARK: convenience properties
+    internal var isActive: Bool {
+        assert(self.eventLoop.inEventLoop)
+        switch self.currentState {
+        case .fresh, .registered, .closed:
+            return false
+        case .activated:
+            return true
+        }
```

Comment:
why not just use?:

```
return self.currentState == .activated
```

Reply:
thanks, same here, used to be more states